admin/services/pupil-status.service.js

```diff
@@ -0,0 +1,145 @@
+const moment = require('moment')
+const R = require('ramda')
+
+const logger = require('./log.service').getLogger()
+const pupilStatusDataService = require('../services/data-access/pupil-status.data.service')
+const pupilIdentificationFlagService = require('./pupil-identification-flag.service')
+const settingService = require('./setting.service')
+const tableSorting = require('../helpers/table-sorting')
+
+const pupilStatusService = {
+  /**
+   * Return the pupil status data
+   * @param schoolId
+   * @return {Promise<*>}
+   */
+  getPupilStatusData: async function (schoolId) {
+    if (!schoolId) {
+      throw new Error('School id not found in session')
+    }
+    const settings = await settingService.get()
+    const pupils = await pupilStatusDataService.sqlFindPupilsFullStatus(schoolId)
+    const pupilsWithStatus = pupils.map(R.partial(pupilStatusService.addStatus, [settings]))
+    return pupilIdentificationFlagService.addIdentificationFlags(tableSorting.applySorting(pupilsWithStatus, 'lastName'))
+  },
+
+  /**
+   * Add the pupil `status` field, using the pupil register service to do so.
+   * @param {Object} settings
+   * @param {Object} pupil
+   * @return {Object} partially cloned pupil obj with an additional `status` property
+   */
+  addStatus: function (settings, pupil) {
+    const newPupil = R.pickAll(['pupilId', 'foreName', 'lastName', 'middleNames', 'dateOfBirth', 'urlSlug', 'checkStatusCode',
+      'group_id', 'reason', 'reasonCode'], pupil)
+    newPupil.status = pupilStatusService.getProcessStatusV2({
+      attendanceId: pupil.attendanceId,
+      checkComplete: pupil.checkComplete,
+      checkReceived: pupil.checkReceived,
+      checkStatusCode: pupil.checkStatusCode,
+      currentCheckId: pupil.currentCheckId,
+      notReceivedExpiryInMinutes: settings.checkTimeLimit,
+      pinExpiresAt: pupil.pinExpiresAt,
+      pupilCheckComplete: pupil.pupilCheckComplete,
+      pupilId: pupil.pupilId,
+      pupilLoginDate: pupil.pupilLoginDate,
+      restartAvailable: pupil.restartAvailable,
+      processingFailed: pupil.processingFailed,
+      reason: pupil.reason
+    })
+    return newPupil
+  },
+
+  /**
+   * Return the process status using new pupil status fields
+   * @param {ProcessStatusArg} arg object containing the parameters
+   * @return {string}
+   */
+  getProcessStatusV2: function (arg) {
+    let status = 'N/A'
+
+    const {
+      attendanceId,
+      checkComplete,
+      checkReceived,
+      checkStatusCode,
+      currentCheckId,
+      notReceivedExpiryInMinutes,
+      pinExpiresAt,
+      pupilCheckComplete,
+      pupilId,
+      pupilLoginDate,
+      restartAvailable,
+      processingFailed,
+      reason
+    } = arg
+
+    if (pinExpiresAt && !moment.isMoment(pinExpiresAt)) {
+      throw new Error('pinExpiresAt must be null or a Moment.moment datetime')
+    }
+
+    if (isPositive(attendanceId)) {
+      status = reason
```

Comment:
This is more of a UI demand from both HDF and pupil status page and it can be moved to a presenter rather than keeping it in the service.
My question is do we maintain the status at that level as `Not taking the Check` or do we allow storing the actual reason?

Reply:
given its shared across both controllers i would leave it as a service, rather than a particular presenter